# list黑盒子

In [1]:
count = 10**5

## 使用append方法

In [6]:
%%timeit
num1 = []
for i in range(count):
    num1.append(i)
num1.reverse()

100 loops, best of 3: 11.6 ms per loop


## 使用insert方法

In [7]:
%%timeit
num2 = []
for i in range(count):
    num2.insert(0,i)

1 loop, best of 3: 3.6 s per loop


比较上面两者的运行时间，可以发现，前者运行11.6ms，后者运行3.6s，两者实现同样的功能，但是时间却相差了3.6 * 1000/11.6=310.344827586倍！

实际上python中的list不是用链表实现的，而是用数组实现的。其主要的设计想法是**将内存分配得过大一些，并且等到其溢出时，在线性时间内再次重新分配内存**。

这样做的效果即是，每次执行insert操作，都需要移动插入点右边的所有元素，知道内存溢出才搬到更大的数组。

而每次执行append操作，只需要在数组后面直接插进去，而不用移动其他元素，除非内存溢出。

实际上，如果我们能确保每次所搬入的数组都是原数组的一定的比例（比如120%或200%），那么将这些搬运开销均摊到每次append操作上的平均成本通常就是常数。

append n个数字的运行时间为O(n)，而在首端insert相同数量的元素的运行时间则为O($n^2$)